quint/src/effects/builtinSignatures.ts

```diff
@@ -241,12 +241,16 @@ const multipleAritySignatures: [QuintBuiltinOpcode, Signature][] = [
   ['tuples', standardPropagation],
   ['and', standardPropagation],
   ['or', standardPropagation],
+  // TODO Effects for match are not correct...
   [
-    'match',
+    'matchVariant',
     (arity: number) => {
-      const readVars = times((arity - 1) / 2, i => `r${i}`)
-      const args = readVars.map(r => `Pure, (Pure) => Read[${r}]`)
-      return parseAndQuantify(`(Read[r], ${args.join(', ')}) => Read[${readVars.join(', ')}]`)
+      const eliminatorIdxs = range((arity - 1) / 2)
+      const vars = eliminatorIdxs.map(i => `Read[r${i}] & Update[u${i}]`)
+      const args = vars.map(vs => `Pure, (${vs}) => ${vs}`)
+      const readVars = ['r', ...eliminatorIdxs.map(i => `r${i}`)].join(', ')
+      const updateVars = ['u', ...eliminatorIdxs.map(i => `r${i}`)].join(', ')
```

Comment:
Hey, I'm not sure what is the problem with the effects you are talking about, but this might be the issue or, at least, part of the issue.
```suggestion
      const updateVars = ['u', ...eliminatorIdxs.map(i => `u${i}`)].join(', ')
```

Reply:
Thanks for taking a look :) I caught this and changed it last week but hadn't pushed the latest to this branch. But I just had a go of writing up how was stuck (with a clear head) and was able to work everything out. Thanks very much! :duck: